### ISSUE TITLE
Get rid of data transfer indirections by having the client directly talk to the storage backend

### ISSUE THREAD

**Issue by derekchiang** [pachyderm/pachyderm#223]
Currently, data goes through three hops before it get persisted: client -> pachd frontend -> pachd backend -> storage servers.

Ideally, we can have clients directly write data to the storage servers, and only the metadata needs to go through pachd.

This will involve an architectural change that's outside of the scope for now; but we should consider it again down the road.

**Comment by JonathanFraser** [comment]
As pachyderm is a versioned system, it is write once read many. A simpler change would be to provide the functionality to retrieve the storage backend key from pachyderm to allow for direct download from the backend.

**Comment by derekchiang** [comment]
@JonathanFraser absolutely.  Pachyderm, being an immutable data store, makes this simple since there can't be race conditions where your data gets modified while you are downloading blocks.

**Comment by jdoliner** [comment]
This should be pretty trivial to do post pfs refactor.

**Comment by derekchiang** [comment]
Bumping the priority of this issue.  We think that by offloading pressure from pachd, we will see less errors related to API timeouts.

**Comment by derekchiang** [comment]
The way we will probably go about doing this is to run a PFS server on the worker node itself, so that it can just talk to localhost to upload/download data.

**Comment by JonathanFraser** [comment]
would it make more sense to split the block storage into a separate gateway process, perhaps running as a daemonset.

**Comment by derekchiang** [comment]
@JonathanFraser I just did some research into `DaemonSet`.  Turns out that there's currently no easy way to discover the daemon running on the local node.  For instance, if a worker is running on node A, then it'd ideally want to use the daemon on node A.  Unfortunately that's not currently possible.  You'd have to put a `Service` on top of a `DaemonSet` and then k8s will route the traffic to a random node, thus defeating the whole point.

See [this issue](https://github.com/kubernetes/kubernetes/issues/24657) for details.  [A PR that exposes host IP](https://github.com/kubernetes/kubernetes/pull/42717) has been merged into master but probably won't see the light until k8s 1.7.
